### PR TITLE
refactor: cache coingecko response

### PR DESCRIPTION
**Pull request by alfonsobries**
## Summary

Closes https://app.clickup.com/t/86drmrzxh

## Checklist

- [ ] I checked that both `pnpm dev` and `pnpm dev:bare` work as intended
- [ ] I checked the basic extension interactions and made sure wallet selection works
- [ ] I checked my UI changes against the design and there are no notable differences, including responsiveness
- [ ] I checked my (code) changes for obvious issues, debug statements and commented code
- [ ] I opened a corresponding card on Clickup for any remaining TODOs in my code
- [ ] I added a short description on how to test this PR _(if necessary)_
- [ ] Ready to be merged

### CODE REVIEW

**Comment by alfonsobries** [comment]
@shahin-hq I completely refactored the hook, I also wasn't considering the maximum number of requests per minute, meaning I was making 1 request per minute instead of N. Could you check again please?

**Review by alexbarnsley — CHANGES_REQUESTED** [review]
I refreshed the extension a few times and eventually got throttled :thinking: if it matters, I'm running 3 mainnet wallets

image

I seem to get the coingecko request on every refresh. It appears that the cache isn't being used after ~1 minute. I've requested changes but if it's user error feel free to dismiss it

File: src/lib/hooks/useExchangeRates.ts [hunk 1]
```diff
@@ -1,26 +1,58 @@
...
+            if (now - ratesCache.lastFetch < ONE_MINUTE_IN_MS) {
+                if (ratesCache.requestCount < MAX_COINGECKO_REQUESTS_PER_MINUTE) {
+                    ratesCache.rates = await fetchRates();
+                    ratesCache.requestCount += 1;
+                } else {
+                    // If MAX_COINGECKO_REQUESTS_PER_MINUTE is reached, use cached rates
+                    setRates(ratesCache.rates);
+                    setLoading(false);
+                    return;
+                }
```
Comment (shahin-hq) [inline src/lib/hooks/useExchangeRates.ts:46]:
```javascript
                if (ratesCache.requestCount < MAX_COINGECKO_REQUESTS_PER_MINUTE) {
                    ratesCache.rates = await fetchRates();
                    ratesCache.requestCount += 1;
                } else {
                    // If MAX_COINGECKO_REQUESTS_PER_MINUTE is reached, use cached rates
                    setRates(ratesCache.rates);
                    setLoading(false);
                    return;
                }
```

What do you think about removing this part?  1 min data is fresh enough I think, I wouldn't keep sending 10 requests in a minute - even it is allowed.

**Comment by alfonsobries** [comment]
@goga-m by using 1 req per minute the logic is going to be simpler so think is not going to be necessary

**Comment by alfonsobries** [comment]
@alexbarnsley the error you had was because a bad merge fixed that and changed the logic to only make one req per minute, should be fine now 

cc @shahin-hq

**Review by shahin-hq — APPROVED** [review]

**Review by alexbarnsley — APPROVED** [review]

**Review by ItsANameToo — APPROVED** [review]

**Pull request merged**